src/dataflow/src/sink/kafka.rs

```diff
@@ -316,6 +320,91 @@ impl KafkaTxProducer {
     }
 }
 
+// So users aren't able to manipulate the fields
+mod consistency_gate {
+    use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+
+    use repr::Timestamp;
+
+    #[derive(Debug)]
+    pub struct KafkaConsistencyGateTs {
+        ts: AtomicU64,
+        init: AtomicBool,
+    }
+
+    impl KafkaConsistencyGateTs {
+        pub fn new() -> Self {
+            Self {
+                ts: AtomicU64::new(0),
+                init: AtomicBool::new(false),
+            }
+        }
+
+        pub fn get_ts(&self) -> Option<Timestamp> {
+            if self.init.load(Ordering::Relaxed) {
```

Comment:
In order for `init` to return `true`, we have to have called `set` -- and if we've called set, then `Some(0)` (the default) is a correct bound -- even if it's not as tight a bound as we'd like `Some(actual_value)`

All that said, I think you're right that `Rc<Cell<Option<Timestamp>>>` is more clear

Reply:
Is the opposite situation also ok? i.e code calls set() and a subsequent get() returns None? Just curious, I agree we should make the code such that it doesn't require this level of analysis

Reply:
Hmm actually I'm starting to doubt where what I just said actually is true because of how I've reordered the code as this PR evolved -- either way, yeah, I'll just side step that question